jdk/src/share/classes/java/util/ServiceLoader.java

```diff
@@ -335,6 +337,7 @@ private class LazyIterator
         ClassLoader loader;
         Enumeration<URL> configs = null;
         Iterator<String> pending = null;
+        Map<String, ClassNotFoundException> loadingExceptions = new HashMap<>();
```

Comment:
The problem is that this check in `hasNextService()` loops through all available names on the first time it's called, not just the next one. So, if multiple exist, single variables are not enough. I can switch the map's value to the class and if not found assume an exception occurred.

Reply:
`hasNextService()` does not need to continue after it has found a class name, for which in the restricted security case `Class.forName()` will have been attempted. That name (or class) will be considered by `nextService()` which will use the previously loaded class, load the class, or fail (because of a `ClassNotFoundException`).

Reply:
Changed the location of the test for allowed or not. Should be fine now.